cms/tests/test_plugins.py

```diff
@@ -1228,6 +1234,51 @@ def test_plugin_toolbar_struct(self):
             self.assertEqual(force_text(style_config['module']), expected_struct_de['module'])
             self.assertEqual(force_text(style_config['name']), expected_struct_de['name'])
 
+    def test_plugin_toolbar_struct_permissions(self):
```

Comment:
The test needs to be done differently because the permission check is done in the rendering call, so we need to test the function output relative to the placeholder rendering.

Here's what I came up with:

```python
from django.template.context import Context

...


def test_plugin_toolbar_struct_permissions(self):
    page = self.get_permissions_test_page()
    staff_user = self.get_staff_user_with_no_permissions()
    placeholder = page.placeholders.get(slot='body')
    page_url = page.get_absolute_url() + '?' + get_cms_setting('CMS_TOOLBAR_URL__EDIT_ON')
    
    self.add_permission(staff_user, 'change_page')
    self.add_permission(staff_user, 'add_text')

    with self.login_user_context(staff_user):
        request = self.get_request(page_url, page=page)
        request.session['cms_edit'] = True
        request.toolbar = CMSToolbar(request)
        content_renderer = self.get_content_renderer(request=request)
        context = {'request': request, 'cms_content_renderer': content_renderer}
        output = content_renderer.render_placeholder(
            placeholder,
            context=Context(context),
            language='en',
            page=page,
            editable=True,
        )
        self.assertIn('<a data-rel="add" href="TextPlugin">Text</a>', output)
        self.assertNotIn('<a data-rel="add" href="LinkPlugin">Link</a>', output)
```